### ISSUE TITLE
imaplib2: 'str' does not support the buffer interface

### ISSUE THREAD

**Issue by nicolas33** [OfflineIMAP/offlineimap#344]
> This v1.0 template stands in `.github/`.
### General informations
- OfflineIMAP version (including imaplib2 if available: `offlineimap -V`): offlineimap v7.0.0-rc3, imaplib2 v2.55 (bundled)
- Python 3.4
### Log error

```
OfflineIMAP 7.0.0-rc3
  Licensed under the GNU GPL v2 or any later version (with an OpenSSL exception)
Account sync z:
 *** Processing account z
 Establishing connection to imap.laposte.net:143
 ERROR: While attempting to sync account 'z'
  'str' does not support the buffer interface
 *** Finished account 'z' in 0:00
ERROR: Exceptions occurred during the run!
ERROR: While attempting to sync account 'z'
  'str' does not support the buffer interface

Traceback:
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/accounts.py", line 273, in syncrunner
    self.__sync()
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/accounts.py", line 336, in __sync
    remoterepos.getfolders()
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/repository/IMAP.py", line 398, in getfolders
    imapobj = self.imapserver.acquireconnection()
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/imapserver.py", line 521, in acquireconnection
    self.__authn_helper(imapobj)
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/imapserver.py", line 422, in __authn_helper
    if func(imapobj):
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/imapserver.py", line 361, in __authn_login
    self.__loginauth(imapobj)
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/imapserver.py", line 203, in __loginauth
    imapobj.login(self.username, self.__getpassword())
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/bundled_imaplib2.py", line 906, in login
    typ, dat = self._simple_command('LOGIN', user, self._quote(password))
  File "/home/nicolas/dev/official_packages/gitrepos/offlineimap/offlineimap/bundled_imaplib2.py", line 1646, in _quote
    return '"%s"' % arg.replace('\\', '\\\\').replace('"', '\\"')
```

**Comment by nicolas33** [comment]
I wonder imaplib2 should shift of type expectation from str to bytes. The library has no way to know which encoding is correct (neither has offlineimap, BTW).

Or, if imaplib2 must keep working with str types, it must encode the strings.

**Comment by nicolas33** [comment]
I expect there are other string manipulations on data we pass to imaplib2. They all have potential issues. Also, string manipulations in imaplib2 on received data from the network might have issues.

Actually, with py3 support we have to handle types and encoding just right. There's no choice but to correctly support unicode.

The best strategy is to decide the encoding and types we want for both the incoming and outgoing data. All of them: configuration file, CLI, filenames, database, UI, etc. Then, it's a matter of policy to decide whether imaplib2 is the "outside" (working with bytes only) or inside (working with unicode only, in which case imaplib2 would have to encode all the strings before sending data out).

Either way, this won't be easy. There are a lot of surprising limitations about encodings: strings in errors, UI libraries, thread names, etc.

I have bad memories about those things. Supporting both py2 and py3 might really be painfull. :-(

**Comment by chris001** [comment]
@nicolas33 imaplib in python 3.5.2 has implemented support for UTF-8 (RFC-6855) in usernames, passwords, email address, and message headers, with the new "enable" method, which outputs "ENABLE UTF8=ACCEPT" to the server.

https://docs.python.org/3/library/imaplib.html
https://tools.ietf.org/html/rfc6855.html

imaplib2 seems to not have UTF-8 RFC-6855 support added, yet...?

we should pull in the UTF-8 RFC-6855 code from imaplib to imaplib2...

**Comment by nicolas33** [comment]
> imaplib2 seems to not have UTF-8 RFC-6855 support added, yet...?

I'd say imaplib2 is designed to be encodings agnostic. However, this can't be done with py3 support in a clean/smart way.

> we should pull in the UTF-8 RFC-6855 code from imaplib to imaplib2...

I don't think that's possible. AFAIK, imaplib2 was a complete rewrite of imaplib.

I don't think the codebase of offlineimap is good enough to play with different encodings now. I don't think offlineimap codebase will ever be good enough to support different encodings properly.

If someone is willing to send patches about that, we'll discuss them and merge them as usual. However, I won't do the job: I've already spent way too many hours on offlineimap in this area in the past. Instead, I intend to go further with imapfw. ,-)

**Comment by chris001** [comment]
@nicolas33 It'd be great if Piers would update impalib2 to support RFC-6855 this way offlineimap could continue to be encodings agnostic.  All that would be necessary would be for imaplib2 to recognize the CAPABILITY when it contains "UTF8=ACCEPT" or "UTF8=ONLY", reply with "ENABLE UTF8=ACCEPT", and communicate the UTF8 strings in quoted-strings and literals....
Not a simple job but Piers probably would instantly see the right way to adapt the new RFC into imaplib2...

**Comment by nicolas33** [comment]
This might help for compatible servers but won't help for the others.
I say "might help" because this won't be easy to properly support this in offlineimap.

**Comment by chris001** [comment]
Ideally we would have the imaplib2 layer (or a new wrapper layer which exists right above it) handle all the nuances of the communication conversation with the IMAP4rev1 servers, including the UTF8 encoding when the server supports RFC-6855.

OfflineImap should only handle the algorithm of the sync, timers, idling, receiving push notifications of message changes and flag changes from the IMAP server.  It shouldn't trouble itself with details of the internals of the IMAP communication layer, to the point where it's taking responsibility for the proper internal operation of the IMAP conversation.  It should take an incoming message as a binary byte stream, don't care the encoding UTF8 or ASCII, and write it to Maildir exactly as it is, no?

There should probably be a separate class to take responsibility to encapsulate the "account config and keyring", i.e., email server name, names of mailboxes to sync, username and authentication keys, etc.

Another class to take charge to read and write the database, to store globally unique keys corresponding to which emails are synced, mailbox sync state, state database locks, etc.

**Comment by chris001** [comment]
Would it be possible to add a switch, in the user config, to specify use imaplib2, or standard python 3.5.2 imaplib.  Has imaplib2 preserved the interface of the original imaplib.  This would solve the issue with the least effort, because the support for the UTF8 is already built into the original imaplib in python 3.5.2.

**Comment by nicolas33** [comment]
I guess that supporting both imaplib and imaplib2 would not be easy. I'm open to merge patches.

**Comment by chris001** [comment]
@nicolas33 Good news, I asked Piers on imaplib2-devel about adding the feature RFC-6855 UTF8 support to imaplib2, and he's adding it now to imaplib2.  He's also fixed some of  the python3 imaplib UTF8 bugs which he's discovered from reading their implementation of RFC-6855...

**Comment by nicolas33** [comment]
Yes, I'm subscribed to imaplib mailing list. The feature was added to the python 3 version of imaplib2 and is still untested. Thanks.

**Comment by thekix** [comment]
Hi,

IMO, this issue should be closed and merged with #670

Cheers,
kix
